### PR TITLE
.Net: MistralAI Function calling for chat completion requests

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Adds support for function calling for non-streaming chat completion requests only.

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs [hunk 1]
```diff
@@ -154,20 +226,38 @@ private MemoryStream GetTestResponseAsStream(string fileName)
         return new MemoryStream(bytes);
     }
 
+    private static HttpRequestHeaders GetDefaultRequestHeaders(string key, bool stream)
+    {
+#pragma warning disable CA2000 // Dispose objects before losing scope
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:231]: Not sure if this pragma is needed once you are returning the reference in the method.

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralChatChoice.cs [hunk 2]
```diff
@@ -15,6 +17,25 @@ internal class MistralChatChoice
...
+    /// The reason the chat completion was finished.
+    /// Enum: "stop" "length" "model_length" "error" "tool_calls"
+    /// </summary>
     [JsonPropertyName("finish_reason")]
     public string? FinishReason { get; set; }
+
+    /// <summary>
+    /// Returns true if the finish reason is "tool_calls"
+    /// </summary>
+    public bool IsToolCall => this.FinishReason?.Equals("tool_calls", StringComparison.Ordinal) ?? false;
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralChatChoice.cs:30]: If we are using this properties just within the connector those can be `internal`

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralPropertyType.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    /// <summary>
+    /// Gets or sets the description of the property.
+    /// </summary>
+    public string? Description { get; set; }
+
+    /// <summary>
+    /// Gets or sets the list of possible values for the property.
+    /// </summary>
+    public IList<string>? Enum { get; set; }
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralPropertyType.cs:25]: Suggest a plural name:

`Enums`? `PossibleValues`? `ValuesDomain` ?

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralToolCall.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel.Connectors.MistralAI.Client;
+internal class MistralToolCall
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralToolCall.cs:6]: Space.

```suggestion

internal class MistralToolCall
```

File: dotnet/src/Connectors/Connectors.MistralAI/MistralAIToolCallBehavior.cs [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            {
+                throw new KernelException($"Auto-invocation with {nameof(EnabledFunctions)} is not supported when no kernel is provided.");
+            }
+
+            foreach (var metadata in this._kernelFunctionMetadata)
+            {
+                // Make sure that if auto-invocation is specified, every enabled function can be found in the kernel.
+                if (autoInvoke)
+                {
+                    Debug.Assert(kernel is not null);
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI/MistralAIToolCallBehavior.cs:199]: Should we have this here?

```suggestion
```

File: dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs [hunk 6]
```diff
@@ -110,4 +112,64 @@ await foreach (var chunk in response)
...
+        };
+        var response = await service.GetChatMessageContentsAsync(chatHistory, executionSettings, kernel);
+
+        // Assert
+        Assert.NotNull(response);
+        Assert.Single(response);
+        Assert.Equal("tool_calls", response[0].Metadata?["FinishReason"]);
+    }
+
+    [Fact] // (Skip = "This test is for manual verification.")
```
Comment (RogerBarreto) [inline dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs:140]: Is intended for Mistral IT to be running in the CI pipeline?

**Review by RogerBarreto — APPROVED** [review]

File: dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs (hunk 6 above)
Comment (markwallace-microsoft) [reply dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs:140]: No, I will skip these when merging to main

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs (hunk 1 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:231]: Without it I get `Severity	Code	Description	Project	File	Line	Suppression State	Details
Warning	CA2000	Call System.IDisposable.Dispose on object created by 'new HttpRequestMessage()' before all references to it are out of scope	Connectors.MistralAI.UnitTests	C:\SemanticKernel\skrepo\semantic-kernel\dotnet\src\Connectors\Connectors.MistralAI.UnitTests\Client\MistralClientTests.cs	231	Active	
`

File: dotnet/src/Connectors/Connectors.MistralAI/MistralAIToolCallBehavior.cs (hunk 5 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.MistralAI/MistralAIToolCallBehavior.cs:199]: Copied from here: https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.OpenAI/ToolCallBehavior.cs#L205

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralPropertyType.cs (hunk 3 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.MistralAI/Client/MistralPropertyType.cs:25]: This class is serialized as a schema. I'm going to replace this in a later PR.

**Pull request merged**